geode-core/src/test/java/org/apache/geode/internal/cache/ClientServerTransactionDUnitTest.java

```diff
@@ -4065,4 +4065,157 @@ private void doJTATx2(String regionName, CountDownLatch latch1, CountDownLatch l
+
+    setupRegionForClientTransactions(totalBuckets, regionName, false, null);
+
+    client.invokeAsync(() -> doKeySetOpTransaction(1, regionName, totalBuckets, false, null));
+
+    // Should cause TXId(server1, 1) to be executed on server2
+    server1.invoke(() -> doPutOpTransaction(regionName, totalBuckets));
+  }
+
+  private void doKeySetOpTransaction(int firstGetKey, String regionName, int totalBuckets,
+      boolean withReplicateRegion, String region2Name) {
+    Region<Integer, String> region = getCache().getRegion(regionName);
+    TXManagerImpl txMgr = (TXManagerImpl) getCache().getCacheTransactionManager();
+    txMgr.begin();
+    region.get(firstGetKey); // starts TXState on a server with the primary bucket of the key
+    verifyKeySetOp(totalBuckets, region);
+
+    if (withReplicateRegion) {
+      Region<Integer, String> region2 = getCache().getRegion(region2Name);
+      int num = totalBuckets + 1;
+      region2.put(num, "" + num);
+      verifyKeySetOp(num, region2);
+    }
+    txMgr.rollback();
+  }
+
+  private void verifyKeySetOp(int expected, Region<Integer, String> region) {
+    Set<Integer> keys = region.keySet();
+    assertEquals(expected, keys.size());
+    for (Integer key : keys) {
+      assertTrue(key <= expected);
+    }
+  }
+
+  private void doPutOpTransaction(String regionName, int totalBuckets) throws InterruptedException {
+    TXManagerImpl txMgr = (TXManagerImpl) getCache().getCacheTransactionManager();
+    Region<Integer, String> region = getCache().getRegion(regionName);
+    txMgr.begin();
+    region.put(2, "NEWVALUE");
+    Thread.currentThread().sleep(100);
+    txMgr.commit();
+  }
+
+  private void doSizeOpTransactions(String regionName, int totalBuckets, String region2Name) {
+    for (int i = 1; i <= totalBuckets; i++) {
+      doSizeOpTransaction(i, regionName, totalBuckets, region2Name);
+    }
+  }
+
+  private void doSizeOpTransaction(int key, String regionName, int totalBuckets,
+      String region2Name) {
+    Region<Integer, String> region = getCache().getRegion(regionName);
+    Region<Integer, String> region2 = getCache().getRegion(region2Name);
+    TXManagerImpl txMgr = (TXManagerImpl) getCache().getCacheTransactionManager();
+    txMgr.begin();
+    region.get(key); // starts TXState on different servers
+    assertEquals(totalBuckets, region.size());
+    int num = totalBuckets + 1;
+    region2.put(num, "" + num);
+    assertEquals(num, region2.size());
+    txMgr.rollback();
+  }
+
+  @Test
+  public void testSizeOpInTransaction() {
+    Host host = Host.getHost(0);
+    VM client = host.getVM(2);
+    String regionName = "region";
+    String region2Name = "region2";
+    int totalBuckets = 2;
+    setupRegionForClientTransactions(totalBuckets, regionName, true, region2Name);
+
+    client.invoke(() -> doSizeOpTransactions(regionName, totalBuckets, region2Name));
+  }
+
+  private void setupRegionForClientTransactions(int totalBuckets, String regionName,
```

Comment:
This method is doing a ton of things (by my count, five different things). Consider breaking down into more methods.

Reply:
Modify the test to use more methods.